cunumeric/fft/fft.py

```diff
@@ -0,0 +1,857 @@
+# Copyright 2021-2022 NVIDIA Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from cunumeric.array import ndarray
+from cunumeric.module import add_boilerplate
+from cunumeric.config import FFTCode, FFTDirection, FFTNormalization
+
+
+def _sanitize_user_axes(a, s, axes, is_c2r=False):
+    if s is None:
+        user_shape = False
+        if axes is None:
+            s = list(a.shape)
+        else:
+            s = [a.shape[ax] for ax in axes]
+    else:
+        user_shape = True
+        s = list(s)
+    if axes is None:
+        axes = list(range(-len(s), 0))
+    if is_c2r and not user_shape:
+        s[-1] = 2 * (a.shape[axes[-1]] - 1)
+    return s, axes
+
+
+def _operate_by_axes(a, axes):
+    return len(axes) != len(set(axes)) or len(axes) != a.ndim or tuple(axes) != tuple(sorted(axes))
```

Comment:
Why does `tuple(axes) != tuple(sorted(axes))` matter? Accuracy-wise, the order of the axis should not really matter (in most cases, not sure about extreme cases), and cuFFT does not guarantees any order anyway I believe. So I think this would not matter? Or is it to simplify the implementation?

Reply:
When axes are specified by the user, we need to differentiate when we can run the problem as a single cuFFT operation, and when we need to run it as multiple 1D FFT over different axes.

For R2C/C2R, numpy does a R2C/C2R 1D FFT over the last axes provided by the user, and then C2C over the remaining axes. cuFFT does R2C/C2R over the innermost (last) axes. So we can only run a normal cuFFT operation if the input axes are sorted. Otherwise, we need to operate by axes to match the numpy behavior.

Granted, this is only for R2C/C2R, so this condition could be made more strict, but I've kept it like this for simplicity for now.